File: client/src/validators.ts
import { z } from "zod";

export const contentItemValidator = z.object({
  id: z.number(),
  date: z.string(),
  title: z.string(),
  description: z.string(),
  url: z.string(),
  media: z
    .array(
      z.object({
        url: z.string(),
        medium: z.string().nullable(),
        type: z.string().nullable(),
        height: z.number().nullable(),
        width: z.number().nullable(),
      }),
    )
    .nullable(),
  source_id: z.number(),
  content_type: z.string(),
});

export const userContentItemValidator = contentItemValidator.extend({
  rating: z.number(),
  source_url: z.string(),
});

export const feedbackValidator = z.object({
  content_id: z.number(),
  rating: z.number(),
});

export const flavourValidator = z.object({
  id: z.number(),
  nickname: z.string().nullable(),
});

File: server/src/validators.py
from datetime import datetime
from pydantic import BaseModel, field_validator
from typing import Optional, Union, List
import json


class Media(BaseModel):
  url: str
  medium: Optional[str] = None  # e.g., "image", "video", etc.
  type: Optional[str] = None  # e.g., "image/jpeg", "video/mp4", etc.
  width: Optional[int] = None
  height: Optional[int] = None


class ContentItem(BaseModel):
  id: int
  title: str
  url: str
  description: str
  media: Optional[List[Media]] = None
  date: datetime
  source_id: int
  content_type: str

  @field_validator("media")
  @classmethod
  def parse_media_json(cls, v):
    if isinstance(v, str):
      try:
        return json.loads(v)
      except json.JSONDecodeError:
        return []
    return v


class UserContentItem(ContentItem):
  rating: int
  source_url: str


class Feedback(BaseModel):
  content_id: int
  rating: int
